libs/x-content/impl/src/test/java/org/elasticsearch/xcontent/provider/json/ESUTF8StreamJsonParserTests.java

```diff
@@ -288,4 +298,116 @@ public void testGetValueRandomized() throws IOException {
         });
     }
 
+    /**
+     * This test compares the retrieval of an optimised text against the baseline.
+     */
+    public void testOptimisedParser() throws Exception {
+        for (int i = 0; i < 200; i++) {
+            String json = randomJsonInput(randomIntBetween(1, 6));
+            try (
+                XContentParser optimisedParser = TestXContentParser.create(json);
+                XContentParser baselineParser = TestXContentParser.create(json)
+            ) {
+                assertThat(optimisedParser.nextToken(), Matchers.equalTo(baselineParser.nextToken()));
+                parseObjectAndAssert(optimisedParser, baselineParser);
```

Comment:
I think this would be easier and should check the baseline and optimized parser produce the same content?

```suggestion
                var expected = baselineParser.mapOrdered();
                var actual = optimisedParser.mapOrdered();
                assertThat(expected, Matchers.equalTo(actual));
```